vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -0,0 +1,662 @@
+  data.shared->l.cb.costs.push_back(data.default_cb_label);
+  for (example* action : data.actions)
+  {
+    action->l.cb.costs = data.cb_label_pool.get_object();
+  }
+}
+
+// the polylabel (union) must be manually cleaned up
+void delete_cb_labels(ccb& data)
+{
+  return_v_array(data.shared->l.cb.costs, data.cb_label_pool);
+
+  for (example* action : data.actions)
+  {
+    return_v_array(action->l.cb.costs, data.cb_label_pool);
+  }
+}
+
+void attach_label_to_example(
+    uint32_t action_index_one_based, example* example, conditional_contexual_bandit_outcome* outcome, ccb& data)
+{
+  // save the cb label
+  // Action is unused in cb
+  data.cb_label.action = action_index_one_based;
+  data.cb_label.probability = outcome->probabilities[0].score;
+  data.cb_label.cost = outcome->cost;
+
+  example->l.cb.costs.push_back(data.cb_label);
+}
+
+void save_action_scores(ccb& data, decision_scores_t& decision_scores)
+{
+  // save a copy
+  auto copy = data.action_score_pool.get_object();
+  copy_array(copy, data.shared->pred.a_s);
+  decision_scores.push_back(copy);
+
+  // correct indices: we want index relative to the original ccb multi-example, with no actions filtered
+  for (auto& action_score : copy)
+  {
+    action_score.action = data.origin_index[action_score.action];
+  }
+
+  // Exclude the chosen action from next slots.
+  auto original_index_of_chosen_action = copy[0].action;
+  data.exclude_list[original_index_of_chosen_action] = true;
+}
+
+void clear_pred_and_label(ccb& data)
+{
+  return_v_array(data.shared->pred.a_s, data.action_score_pool);
+
+  // This just needs to be cleared as it is reused.
+  data.actions[data.action_with_label]->l.cb.costs.clear();
+}
+
+// true if there exists at least 1 action in the cb multi-example
+bool has_action(multi_ex& cb_ex) { return cb_ex.size() > 1; }
+
+// This function intentionally does not handle increasing the num_features of the example because
+// the output_example function has special logic to ensure the number of feaures is correctly calculated.
+// Copy anything in default namespace for slot to ccb_slot_namespace in shared
+// Copy other slot namespaces to shared
+void inject_slot_features(example* shared, example* slot)
+{
+  for (auto index : slot->indices)
+  {
+    // constant namespace should be ignored, as it already exists and we don't want to double it up.
+    if (index == constant_namespace)
+    {
+      continue;
+    }
+    else if (index == default_namespace)  // slot default namespace has a special namespace in shared
+    {
+      LabelDict::add_example_namespace(*shared, ccb_slot_namespace, slot->feature_space[32]);
+    }
+    else
+    {
+      LabelDict::add_example_namespace(*shared, index, slot->feature_space[index]);
+    }
+  }
+}
+
+template <bool audit>
+void inject_slot_id(ccb& data, example* shared, int id)
+{
+  // id is zero based, so the vector must be of size id + 1
+  if(id + 1 > data.hashes.size())
+  {
+    data.hashes.resize(id + 1, 0);
```

Comment:
>hashes [](start = 9, length = 6)

If this is only used to store slotid hashes, rename to slotid_hashes.  It's nice that this needs to be only populated once.